### PR TITLE
Add Tests w/ 100% Code Coverage

### LINKED ISSUE #4: Add tests
Noting that I'll likely need to mock out `localStorage` for tests. Not difficult given that a simple mock can be < 10 LoC (or use some in-memory package).

Unlike Jest, AVA doesn't support mocking out-of-the-box, but we don't really need `sinon` or too much mocking support for a small mock like this, so should be nbd

### PR DESCRIPTION
**Pull request by agilgur5**
## Pull Request

- create test harness
- ensure basic persist functionality works
- ensure persist options work
- add CodeCov test coverage reporting

Fixes #4 

## Process to get to completion

Was originally planning to get this done last week, but immediately hit bugs when I got `Cannot find name "describe"` TypeScript errors. Adding `@types/jest` did not resolve the issue. Earlier today I fixed the VSCode TS errors by adding an `import @types/jest` to the top, but then `ts-jest` errored out on that. Eventually fixed both by adding a `/// <reference types="@types/jest" />` at the top instead.
I'm not sure why a reference declaration is necessary if TS is supposed to auto-import from `@types`... Every SO Q&A I see says it should work past TS 2.1, and I'm using TS ^3.5.2...

I also ran into another issue immediately after that, which was that I couldn't get tests to run from a `tests/` directory because apparently Jest only checks the `test/` (singular) directory 😕 

Was mostly smooth sailing writing tests after that, especially as `jsdom` actually has a `localStorage` implementation built-in (and `jest` has a  `jsdom` integration built-in... though it would probably make for a fatest env/faster tests to just polyfill `localStorage` in isolation).

I did run into #18 while using the `localStorage` implementation immediately when I tried running the tests, so I actually had to fix that in the process 😅 . See the PR to fix that in #22  
Also added a test for #19 in this PR as well.
So that means this PR is dependent on #22 and #19 being merged before it can be merged. Locally I got all the tests to run, but this PR is split off to only have the tests related code.

When I got to adding code coverage, I noticed the AsyncLocalStorage `clear` and `removeItem` functions were causing a lot of lost coverage, but they're actually not used internally, they're more for consistency and typing (and might later be exposed externally). Had to update TSDX to get [`jest.config.js` support](https://github.com/jaredpalmer/tsdx/releases/tag/v0.10.0) so that I could add an ignore for `asyncLocalStorage.ts`. Was confused for a bit why my `jest.config.js` didn't work as I saw it was supported, but apparently this was a fairly recent feature.
Then decided not to ignore the whole file, but just the two functions instead so that only the parts that need to be ignored are. Ran into some issues with ignoring coverage for ES6 class functions that seem related to https://github.com/gotwarlost/istanbul/issues/445 (and I guess the istanbul org version still has the same issue -- the old repo should really be archived to avoid confusion). The workarounds listed there fixed it, but required some awkwardly placed comments.
*EDIT*: output definitely seems related to that issue. the awkwardly placed comments result in:
```javascript
  clear
  /* istanbul ignore next */
  : function clear() {
```
so I guess that makes sense why it has to go in between the name and parens then until this functionality is fixed.

## Future Work

- [ ] Possibly add functionality to throw an error if `jsonify` is set to `false` and `localStorage` is the storage engine chosen? (bc it will save as `'[object Object]'` and that's almost certainly not intentional usage). We don't generally throw on improper usage of the Storage Engine though, and just put the responsibility on the user to know how their chosen engine works (e.g. could similarly error if `AsyncStorage` is chosen and `jsonify` is `false` -- that starts digging a rabbit hole of checking all sorts of engines. Also would need to know how to check if which storage it is without importing that storage). On the other hand, `localStorage` is the one and only default. 🤷‍♂ 
- [ ] [Add an integration test for `localForage`](https://github.com/agilgur5/mst-persist/issues/4#issuecomment-510624604) and possibly other storage engines, like maybe `AsyncStorage` and at least one `redux-persist`-specific storage engine to ensure compatibility.
- [x] ~Still missing coverage of the two `Promise.reject` in `index.ts` and in `asyncLocalStorage.ts`. They're both kind of tricky edge cases to test, so maybe I should ignore them? The former requires writing a test to run specifically in `jest --env=node` instead of `jsdom` (which means doing some hackish code to split where or when that test runs) and the latter requires doing an invalid operation on `AsyncLocalStorage`, which shouldn't happen in internal usage (so that would be part of a potential future AsyncLocalStorage unit test suite, so maybe it should indeed be ignored like `clear` and `removeItem`)~.
**EDIT**: `Promise.reject` in `asyncLocalStorage.ts` has been ignored similar to `clear` and `removeItem` per the reasoning above that it's not part of internal usage.
**EDIT2**: A node test has been added for the edge case of `Promise.reject` in `index.ts`.

### CODE REVIEW

**Comment by agilgur5** [comment]
Updated the PR to ignore the `catch` / `Promise.reject` in `asyncLocalStorage.ts` as this part, like I reasoned out while I was writing up the "Future Work" section above, is also not used internally (i.e. changing calls to `persist` etc will not cause `clear`, `removeItem`, `Promise.reject` to trigger as no internal code uses those).

~A `jest --env=node` test should still be added at some point, as that is valid internal usage, so not quite at 100% coverage yet.~ **EDIT**: added it in this PR and got to 100% coverage!

I also changed the `[object Object]` comparison to a `toBe` as `toStrictEqual` is not necessary in this case as they're both just strings (and not objects like the rest).

And for some reason Travis CI checks aren't appearing in the PR GitHub UI right now (possibly because of bugs with general release of Actions?), so had to manually check to see that Node v8.9  or NPM v5.6 was causing some CI failures. Upgraded to Node v10 Active LTS and problems solved and CI passes -- all green!

**Comment by agilgur5** [comment]
Decided I might as well see what I can do about adding a Node test, and turns out it's actually not as complicated as I thought because Jest does support changing [`jest-environment` per test file](https://jestjs.io/docs/en/configuration#testenvironment-string). Added the test and got to 100% code coverage!

I would've preferred having the Node test in the same file as the other tests, but unfortunately Jest doesn't support setting `jest-environment` on a per test basis, only per file. The PR that implemented the per file support is actually based on [an issue](https://github.com/facebook/jest/issues/2587) that feature requested per test changes, but that seems to have never gotten implemented _all the way_ to per test.

File: test/index.spec.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  })
+
+  it('should whitelist', async () => {
+    const user = UserStoreF.create()
+    await persist('user', user, {
+      whitelist: ['name']
+    })
+
+    user.changeName('Joe') // fire action to trigger onSnapshot
+    const snapshot = { ...getSnapshot(user) }
```
Comment (agilgur5) [inline test/index.spec.ts]: Might want to comment why this is necessary here and in the actual source code. I didn't know until now why it was necessary (as I had borrowed that piece from a gist). 

It throws a [`TypeError: Cannot delete property...`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Non_configurable_array_element) when trying to `delete` a key otherwise as the properties have apparently been set with `Object.defineProperty`.

**Review by agilgur5 — COMMENTED** [review]
Has gone through multiple iterations already and looks 99% good! Just one tiny comment I should probably add

**Review by agilgur5 — COMMENTED** [review]
seems all good to go finally!
